File: cole/day07_binary-search/binary-search.c
// faster 72% solution

int search(int *nums, int numsSize, int target)
{
    int *lo = nums, *hi = nums + numsSize - 1;
    if (*lo == target)
        return 0;
    if (*hi == target)
        return numsSize - 1;
    for (int *med; lo != hi;)
    {
        med = lo + ((hi - lo) / 2);
        if (*med == target)
            return (med - nums);
        if (*med < target)
            lo = med + 1;
        else
            hi = med;
    }
    return (*lo == target) ? (lo - nums) : -1;
}

// simple, "correct" solution

int search(int *nums, int numsSize, int target)
{
    int lo = 0, hi = numsSize - 1;
    for (int med; lo != hi;)
    {
        med = (hi + lo) / 2;
        if (nums[med] < target)
            lo = med + 1;
        else
            hi = med;
    }
    return (nums[lo] == target) ? lo : -1;
}

File: cole/day07_binary-search/binary-search.py
# the "fuck you it works" way
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        return nums.index(target) if target in nums else -1


# the "proper" way
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        lo, hi = 0, len(nums) - 1
        if nums[0] == target:
            return nums[0]
        if nums[-1] == target:
            return len(nums) - 1
        mid = len(nums) // 2
        while lo != hi and nums[mid] != target:
            if nums[mid] < target:
                lo = mid + 1
            else:
                hi = mid
            mid = (lo + hi) // 2
        return mid if nums[mid] == target else -1
